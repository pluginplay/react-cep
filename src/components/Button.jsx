import styled from 'styled-components'

export default styled.button`
  padding: 4px 8px;
  margin: 10px auto;
  min-width: 100px;
  font-family: "${props => props.theme.ae.baseFontFamily}", "Segoe UI", "San Francisco", sans-serif;
  color: white;
  background: ${props => props.theme.background.dark};
  border: solid 1px ${props => props.theme.background.light};
  font-size: ${props => props.theme.ae.baseFontSize}pt;
  outline: none;
  box-sizing: border-box;
  transition: background 0.1s ease-in-out, border-color 0.1s ease-in-out, transform 0.1s ease-in-out;
  transform: scale(1);
  cursor: pointer;
  display: inline-block;
  &:hover, &:focus, &:active {
    border-color: ${props => props.theme.ae.systemHighlightColor};
    background: ${props => props.theme.background.darker};
  }
  &:active { 
    transform: scale(0.95);
    background: ${props => props.theme.ae.panelBackgroundColor};
  }
`
