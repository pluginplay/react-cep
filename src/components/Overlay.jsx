import Flex from 'styled-flex-component'

export default Flex.extend`
  transition: opacity 0.25s linear, z-index 0s linear ${props => props.visible ? '' : '0.25s'};
  text-align: center;
  position: absolute;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  background: ${props => props.theme.background.dark};
  z-index: ${props => props.visible ? 3 : -1};
  opacity: ${props => props.visible ? 1 : 0};
  a {
    color: white;
    text-decoration: none;
    opacity: 0.75;
    transition: opacity 0.2s linear;
    &:hover {
      opacity: 1;
    }
  }
`
