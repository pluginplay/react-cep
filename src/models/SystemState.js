import { observable, action } from 'mobx'

export default class SystemState {
  @observable progress = 0
  @observable progressLabel = 'Loading'
  @observable showProgress = false
  @observable showProgressComplete = false
  @observable errorMessage = null
  @observable errorObject = null
  @observable showErrorOverlay = false
  @observable debugMessage = null
  @observable showDebugOverlay = null
  @observable confirmationMessage = null
  @observable confirmationCallback = null
  @observable showConfirmationOverlay = false

  @action setProgress (progress, label, showProgress = true) {
    this.showProgressComplete = false
    this.showProgress = showProgress
    this.progressLabel = label
    this.progress = progress
  }

  @action progressComplete () {
    this.progress = 100
    this.showProgressComplete = true
    setTimeout(() => {
      this.hideProgress()
    }, 3000)
  }

  @action hideProgress () {
    this.showProgress = false
  }

  @action setErrorMessage (errorText, error) {
    this.errorMessage = errorText
    this.errorObject = error
    this.showErrorOverlay = true
  }

  @action hideErrorMessage () {
    this.errorMessage = ''
    this.errorObject = null
    this.showErrorOverlay = false
  }

  @action addDebugMessage (message) {
    console.log(`[DEBUG] ${message}`)
    if (this.debugMessage === null) {
      this.debugMessage = message
    } else {
      this.debugMessage += `\n${message}`
    }
  }

  @action showDebug () {
    this.showDebugOverlay = true
  }

  @action clearDebug () {
    this.showDebugOverlay = false
    this.debugMessage = null
  }

  @action showConfirmation (message, callback) {
    this.showConfirmationOverlay = true
    this.confirmationMessage = message
    this.confirmationCallback = callback
  }

  @action hideConfirmation () {
    this.showConfirmationOverlay = false
    this.confirmationMessage = null
    this.confirmationCallback = null
  }
}
