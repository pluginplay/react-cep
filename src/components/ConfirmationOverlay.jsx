import React from 'react'
import SystemState from '../models/SystemState'
import PropTypes from 'prop-types'
import Overlay from './Overlay'
import OverlayContent from './OverlayContent'
import Button from './Button'
import { observer } from 'mobx-react'

@observer
class ConfirmationOverlay extends React.Component {
  constructor (props) {
    super(props)
    this._onCloseClickedBound = this.onCloseClicked.bind(this)
  }

  onCloseClicked () {
    if (this.props.systemState.confirmationCallback) {
      this.props.systemState.confirmationCallback()
    }
    this.props.systemState.hideConfirmation()
  }

  render () {
    const visible = this.props.systemState.showConfirmationOverlay
    const message = this.props.systemState.confirmationMessage
    return (
      <Overlay visible={visible} center column>
        <OverlayContent column>
          <div>{message}</div>
          <Button onClick={this._onCloseClickedBound}>OK</Button>
        </OverlayContent>
      </Overlay>
    )
  }
}

ConfirmationOverlay.propTypes = {
  systemState: PropTypes.instanceOf(SystemState).isRequired
}

export default ConfirmationOverlay
