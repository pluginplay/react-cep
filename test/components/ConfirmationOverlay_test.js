import React from 'react'
import { shallow } from 'enzyme'
import { expect, withTheme } from '../setup'
import renderer from 'react-test-renderer'
import SystemState from '../../src/components/SystemState'
import ConfirmationOverlay from '../../src/components/ConfirmationOverlay'

import Overlay from '../../src/components/Overlay'
import Button from '../../src/components/Button'

let systemState = null
let hideConfirmationSpy = null

beforeEach(() => {
  hideConfirmationSpy = jest.spyOn(SystemState.prototype, 'hideConfirmation')
  systemState = new SystemState()
  systemState.showConfirmationOverlay = true
  systemState.confirmationMessage = 'Hello world?'
})

afterEach(() => {
  hideConfirmationSpy.mockRestore()
})

const getWrapper = () => {
  return shallow(<ConfirmationOverlay systemState={systemState} />)
}

it('is only visible when showConfirmationOverlay is true', () => {
  const wrapper = getWrapper()
  expect(wrapper.find(Overlay).first()).to.have.prop('visible').equal(true)
})

it('displays the proper message', () => {
  const wrapper = getWrapper()
  expect(wrapper.find('div')).to.have.text('Hello world?')
})

it('calls hideConfirmation when the button is clicked', () => {
  const wrapper = getWrapper()
  wrapper.find(Button).first().simulate('click')
  jestExpect(hideConfirmationSpy).toHaveBeenCalledTimes(1)
})

it('calls systemState.confirmationCallback when the button is clicked', () => {
  const confirmationCallback = jest.fn()
  systemState.confirmationCallback = confirmationCallback
  const wrapper = getWrapper()
  wrapper.find(Button).first().simulate('click')
  jestExpect(confirmationCallback).toHaveBeenCalledTimes(1)
})

it('renders the proper HTML structure', () => {
  const tree = renderer
    .create(withTheme(<ConfirmationOverlay systemState={systemState} />))
    .toJSON()
  jestExpect(tree).toMatchSnapshot()
})
