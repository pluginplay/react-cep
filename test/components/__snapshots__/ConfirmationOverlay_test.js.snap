// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the proper HTML structure 1`] = `
<div
  className="flex-bfOkaE fdXOYB"
>
  <div
    className="flex-hMftNG iHsMmk"
  >
    <div>
      Hello world?
    </div>
    <button
      className="sc-bdVaJa jOdvgS"
      onClick={[Function]}
    >
      OK
    </button>
  </div>
</div>
`;
